Enable test_torch.py tests for BFloat16 on cuda

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22605 Added .bfloat16()
* **#22428 Enable test_torch.py tests for BFloat16 on cuda**

Differential Revision: [D16083575](https://our.internmc.facebook.com/intern/diff/D16083575)

izdeby committed:
BFloat16 for cuda

izdeby committed:
Update on "[WIP] BFloat16 for cuda"

BFloat16 for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

izdeby committed:
Update on "[WIP] BFloat16 for cuda"

BFloat16 for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

izdeby committed:
Update on "Enable TestTorch tests for BFloat16 on cuda"

BFloat16 for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

ezyang commented on aten/src/ATen/preprocess_declarations.py:
```diff
@@ -76,9 +76,9 @@ def expand(types):
-        backend_types['CUDA'].discard('BFloat16')
+    if not option.get('cuda_bfloat16', False):
+        if 'CUDA' in backend_types:
+            backend_types['CUDA'].discard('BFloat16')
```
Not really removed, eh? :)

izdeby replied:
its coming soon :)

ezyang commented:
It's a pretty big patch but it all looks reasonable to me. Letting @gchanan take a look.

izdeby commented:
> It's a pretty big patch but it all looks reasonable to me. Letting @gchanan take a look.

@ezyang, yeah i know its big. The problem is that there are a lot of inner dependencies from one thing to another. For example, i had to enable couple methods just to make test utils work with BFloat16 type as it is and without hacks.

izdeby committed:
Update on "Enable TestTorch tests for BFloat16 on cuda"

BFloat16 for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

izdeby committed:
Update on "Enable TestTorch tests for BFloat16 on cuda"

≈BFloat16 for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

izdeby commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -6,6 +6,7 @@
   cpu_bool: True
   cuda_bool: True
   cpu_bfloat16: True
+  cuda_bfloat16: True
```
This PR is already big enough. More functionality will come in next PRs.

izdeby commented on aten/src/ATen/native/cuda/BinaryOpsKernel.cu:
```diff
@@ -13,7 +13,7 @@
 
 void add_kernel_cuda(TensorIterator& iter, Scalar alpha_scalar) {
-  AT_DISPATCH_ALL_TYPES_AND(kHalf, iter.dtype(), "add_cuda", [&]() {
+  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "add_cuda", [&]() {
```
This is needed for the test/common_utils.py. The way we compare two tensors requires some math ops.

izdeby committed:
Update on "Enable test_torch.py tests for BFloat16 on cuda"

Enable test_torch BFloat16 tests for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

izdeby committed:
Update on "Enable test_torch.py tests for BFloat16 on cuda"

Enable test_torch BFloat16 tests for cuda

gh-metadata: pytorch pytorch 22428 gh/izdeby/13/head

izdeby commented on aten/src/THC/CMakeLists.txt:
```diff
@@ -26,6 +26,14 @@ foreach(THC_FILE TensorMathCompareT TensorMathCompare TensorMathReduce TensorMas
    LIST(APPEND extra_src "${CMAKE_CURRENT_SOURCE_DIR}/generated/THC${THC_FILE}Bool.cu")
 endforeach()
 
+foreach(THC_FILE TensorMathReduce TensorMathCompareT TensorMathCompare TensorMathPointwise)
```
sorted
